lib/datadog/security/contrib/rack/request_middleware.rb

```diff
@@ -0,0 +1,93 @@
+
+            libddwaf_spec = Gem.loaded_specs['libddwaf']
+            libddwaf_platform = libddwaf_spec ? libddwaf_spec.platform.to_s : 'unknown'
+            ruby_platforms = Gem.platforms.map(&:to_s)
+
+            if libddwaf_required?
+              Datadog.logger.debug { "libddwaf platform: #{libddwaf_platform}" }
+              Datadog::Security::WAF.logger = Datadog.logger if Datadog.logger.debug?
+              @waf = Datadog::Security::WAF::Handle.new(waf_rules)
+
+              # TODO: check is too low level
+              # TODO: use proper exception
+              raise if @waf.handle_obj.null?
+            else
+              Datadog.logger.warn { "libddwaf is missing components. installed platform: #{libddwaf_platform} ruby platforms: #{ruby_platforms}" }
+              Datadog.logger.warn { 'AppSec is disabled' }
+            end
+          end
+
+          def waf_rules
+            @waf_rules ||= JSON.parse(Datadog::Security::Assets.waf_rules)
+          end
+
+          def call(env)
+            return @app.call(env) unless libddwaf_required?
+
+            # TODO: handle exceptions, except for @app.call
+
+            context = Datadog::Security::WAF::Context.new(@waf)
+            # TODO: check is too low level
+            # TODO: use proper exception
+            raise if context.context_obj.null?
+
+            env['datadog.waf.context'] = context
+            request = ::Rack::Request.new(env)
+
+            request_return, request_response = Instrumentation.gateway.push('rack.request', request) do
+              @app.call(env)
+            end
+
+            if request_response && request_response.any? { |action, _event| action == :block }
+              request_return = [403, { 'Content-Type' => 'text/html' }, [Datadog::Security::Assets.blocked]]
+            end
+
+            response = ::Rack::Response.new(request_return[2], request_return[0], request_return[1])
+            response.instance_eval do
+              @waf_context = context
+            end
+
+            _, response_response = Instrumentation.gateway.push('rack.response', response)
+
+            request_response.each { |_, e| e.merge!(response: response) } if request_response
+            response_response.each { |_, e| e.merge!(request: request) } if response_response
+            both_response = (request_response || []) + (response_response || [])
+
+            if both_response.any?
+              Security::Event.record(*both_response.map { |_action, event| event })
+            end
+
+            request_return
+          end
+
+          def libddwaf_required?
+            defined?(Datadog::Security::WAF)
+          end
+
+          def require_libddwaf
+            require 'libddwaf'
+          rescue LoadError => e
+            Datadog.logger.warn { "LoadError: libddwaf failed to load: #{e.message}. Try adding `gem 'libddwaf'` to your Gemfile" }
+            Datadog.logger.warn { 'Appsec is disabled' }
+          end
```

Comment:
Is this dynamic loading of libddwaf still needed? Since it's being added as a regular dependency?

Reply:
Yes, for example libddwaf does not support JRuby yet. Also this makes it so that non-appsec users are guaranteed to not suffer from any hypothetical issue or process size increase with the shared lib being loaded.

Reply:
> Yes, for example libddwaf does not support JRuby yet.

To clarify, the `ruby` platform libddwaf is not an empty shim, it just fails to load the relevant shared lib for currently unsupported platforms.

Reply:
Ah... I believe I was misled by the error message. Can we change/improve it? For instance, can we tell users "this doesn't work on JRuby", rather than the current suggestion of trying to add it as a dependency which actually wouldn't fix the issue?

We do something similar for the profiler as well:

* https://github.com/DataDog/dd-trace-rb/blob/6b0b1d51a4a845339a6b986841f66077034c3e6b/lib/ddtrace/profiling.rb#L18
* https://github.com/DataDog/dd-trace-rb/blob/6b0b1d51a4a845339a6b986841f66077034c3e6b/lib/ddtrace/profiling/tasks/setup.rb#L42